### PR TITLE
Load constants lazily on RailsAdmin initialization

### PR DESCRIPTION
**Pull request by mshibuya**
This is a counter-proposal for #3492, and aims to fix issues #3490, #3501, #3533, and #3534.

My apologies for those issues. This is because of the complexity in RailsAdmin (and Rails) initialization process, which is hard to get it right. Let me explain the story...

RailsAdmin initializer (`config/initializers/rails_admin.rb`) is loaded during the Rails app's startup process. But as for the actual configuration (= execution of the DSL itself, the code passed in the `RailsAdmin.config do |config| ... end` block), there're many constraints that all need to be fulfilled for a successful initialization.

- At this point, basically dependent gems are already loaded but the application codes (e.g. files under `app/` directory) are not loaded yet.
- But historically RailsAdmin [allows model constants to be passed as the `config.model` argument](https://github.com/railsadminteam/rails_admin/pull/3492#issuecomment-1064915362). This led to issues like #3025 in pre-Zeitwerk age, and with Zeitwerk it's not possible because its autoload feature is not enabled at this moment.

So the attempt to address this was made in [e4ae669](https://github.com/railsadminteam/rails_admin/commit/e4ae6698e52e56a1cefdf5c4097b29b8306f21e2). The idea was basically to delay the execution of the configuration block, after model classes are ready to be autoloaded. But it brought different problems:

- Devise [performs `reload_routes!` on `before_eager_load`](https://github.com/heartcombo/devise/blob/8593801130f2df94a50863b5db535c272b00efe1/lib/devise/rails.rb#L17). That means RailsAdmin actions ([which are used to construct RailsAdmin routes](https://github.com/railsadminteam/rails_admin/blob/master/config/routes.rb)) need to be set up ahead of that, unless RailsAdmin routes fail to be reflected. To work aronud this issue, `reload_routes!` was [also added](https://github.com/railsadminteam/rails_admin/commit/08f50aabc1922aeb289ced4ccbe9f983bc1aaa89) in the RailsAdmin initialization process. And that solution isn't perfect, resulting in #3533.
- Upon the introduction of [the Importmap support](https://github.com/railsadminteam/rails_admin/pull/3488), [RailsAdmin's asset configuration](https://github.com/railsadminteam/rails_admin/blob/e59a9353e742b0f6e6b152d1b1f9e7cafad3ee54/lib/rails_admin/engine.rb#L45) must happen earlier than `after_initialize` (at that point, it's too late to modify values in `app.config.assets`). So [the configuration execution](https://github.com/railsadminteam/rails_admin/blob/e59a9353e742b0f6e6b152d1b1f9e7cafad3ee54/lib/rails_admin/engine.rb#L38) was brought earlier, but that didn't go well and brought #3534.

Thus, it's like 2 contradicting forces pulling the initialization timing both forward ant backward. That's why I have been struggling in solving this issue without success...

We need a new way to overcome this problem, and here I have come up with one. That is to intercept `const_missing` to eliminate the necessity for loading actual classes, while delaying the model initialization by utilizing `LazyModel`, which is the proven method from pre RailsAdmin 3.x era.
At the first sight overriding `Object.const_missing` seemed to have side effects, but from what I tested it looks working well. So I want you guys to try this in your setup, and give feedback on whether this is good idea or not.

Thank you for reading this through, and hoping to hearing back!

@codealchemy @jdufresne @q3aiml (or anyone who is interested in this)

### CODE REVIEW

File: spec/dummy_app/config/initializers/rails_admin.rb [hunk 1]
```diff
@@ -2,8 +2,8 @@
 
 RailsAdmin.config do |c|
   c.asset_source = CI_ASSET
-  c.model 'Team' do
```
Comment (codealchemy) [inline spec/dummy_app/config/initializers/rails_admin.rb:5]: I think this was where I got tripped up in my config (using model constants in the initializer), leading to https://github.com/railsadminteam/rails_admin/issues/3534

It feels problematic to continue to allow this (and makes sense to me that one shouldn't reference model _constants_ in an initializer, https://github.com/railsadminteam/rails_admin/issues/3542 highlights another possibly related issue). If need be, the config block for a model can already live in the model itself (which helps keep the initializer more manageable in large applications).

I'd be in favor of removing support for using model constants in the initializer entirely (given that using the model name as a String continues to work as expected), particularly if that can help avoid needing to override `const_missing`

**Review by codealchemy — COMMENTED** [review]
I can confirm that this change addresses the issues I was running into in https://github.com/railsadminteam/rails_admin/issues/3534

Some thoughts re: allowing model constants to be used in the initializer noted above.

File: spec/dummy_app/config/initializers/rails_admin.rb (hunk 1 above)
Comment (codealchemy) [reply spec/dummy_app/config/initializers/rails_admin.rb:5]: Got it, I think it works well given the cases needed to be handled here - just always cautious with overrides

**Review by codealchemy — APPROVED** [review]

**Comment by mshibuya** [comment]
Merged. Thank you for reviewing!

**Comment by q3aiml** [comment]
This is working well in my testing and resolves https://github.com/railsadminteam/rails_admin/issues/3490. Thank you for figuring this out!

**Pull request merged**